[DebugInfo] Add num_extra_inhabitants to debug info

augusto2112 opened the pull request:
An extra inhabitant is a bit pattern that does not represent a valid
value for instances of a given type. The number of extra inhabitants is
the number of those bit configurations.

This is used by Swift to save space when composing types. For example,
because Bool only needs 2 bit patterns to represent all of its values
(true and false), an Optional<Bool> only occupies 1 byte in memory by
using a bit configuration that is unused by Bool. Which bit patterns are
unused are part of the ABI of the language.

Since Swift generics are not monomorphized, by using dynamic libraries
you can have generic types whose size, alignment, etc, are known only
at runtime (which is why this feature is needed).

This patch adds num_extra_inhabitants to LLVM-IR debug info and in DWARF
as an Apple extension.

pogo59 reviewed (commented):
A count of invalid bit patterns doesn't seem that helpful to the debugger. What would it do with this information? How would it know what the actual invalid bit patterns are? For example, an enum with two enumerators with values equal to 0 and 5 would have exactly the same number of "extra inhabitants" as bool.

augusto2112 commented:
> What would it do with this information? 

This information is used to calculate the layout of types. For example, since Bool has extra inhabitants, the debugger needs that information to calculate that Optional<Bool> will only use 1 byte.

Since Swift generics are not monomorphized, by using dynamic libraries you can have generic types whose size, alignment, etc, are known only at runtime (which is why this feature is needed).

> How would it know what the actual invalid bit patterns are? For example, an enum with two enumerators with values equal to 0 and 5 would have exactly the same number of "extra inhabitants" as bool.

This is part of the ABI of the language (which is why I added the attribute as an Apple extension). You are correct that if you have an enum with two enumerators equal to 0 and 5 you can't know which bit patterns are extra inhabitants. But for enums that don't specify what the integer representation is, two enumerators means that all bit patterns outside 0 and 1 are extra inhabitants.

dwblaikie commented:
Can you include some of these details in the description/commit message? 
(this bit, or something like it, especially: "Since Swift generics are not monomorphized, by using dynamic libraries you can have generic types whose size, alignment, etc, are known only at runtime (which is why this feature is needed).")

augusto2112 commented:
> Can you include some of these details in the description/commit message? (this bit, or something like it, especially: "Since Swift generics are not monomorphized, by using dynamic libraries you can have generic types whose size, alignment, etc, are known only at runtime (which is why this feature is needed).")

Ok, I updated the commit message.

augusto2112 committed:
[DebugInfo] Add num_extra_inhabitants to debug info

An extra inhabitant is a bit pattern that does not represent a valid
value for instances of a given type. The number of extra inhabitants is
the number of those bit configurations.

This is used by Swift to save space when composing types. For example,
because Bool only needs 2 bit patterns to represent all of its values
(true and false), an Optional<Bool> only occupies 1 byte in memory by
using a bit configuration that is unused by Bool. Which bit patterns are
unused are part of the ABI of the language.

Since Swift generics are not monomorphized, by using dynamic libraries
you can have generic types whose size, alignment, etc, are known only
at runtime (which is why this feature is needed).

This patch adds num_extra_inhabitants to LLVM-IR debug info and in DWARF
as an Apple extension.

augusto2112 commented:
> Oh, do you have a link to the ABI documentation that describes/shows how "num extra inhabitants" would be enough to know which bit patterns are available for extra inhabitants?

There's _some_ documentation at https://github.com/swiftlang/swift/blob/main/docs/ABIStabilityManifesto.md and https://github.com/swiftlang/swift/blob/main/docs/ABI/TypeLayout.rst, although it doesn't seem to be complete, unfortunately.

adrian-prantl commented on llvm/include/llvm/IR/DIBuilder.h:
```diff
@@ -225,9 +225,13 @@ namespace llvm {
     /// \param Flags       Optional DWARF attributes, e.g., DW_AT_endianity.
+    /// \param NumExtraInhabitants The number of extra inhabitants of the type.
+    /// An extra inhabitant is a bit pattern that does not represent a valid
+    /// value for instances of a given type.
```
Can you add a comment that this is used by the Swift language?

adrian-prantl commented on llvm/include/llvm/BinaryFormat/Dwarf.def:
```diff
@@ -636,6 +636,7 @@ HANDLE_DW_AT(0x3fed, APPLE_property, 0, APPLE)
 HANDLE_DW_AT(0x3fee, APPLE_objc_direct, 0, APPLE)
 HANDLE_DW_AT(0x3fef, APPLE_sdk, 0, APPLE)
 HANDLE_DW_AT(0x3ff0, APPLE_origin, 0, APPLE)
+HANDLE_DW_AT(0x3ff1, APPLE_num_extra_inhabitants, 0, APPLE)
```
I would suggest moving this out of the APPLE namespace into the LLVM one.

adrian-prantl commented on llvm/include/llvm/IR/DebugInfoMetadata.h:
```diff
@@ -712,31 +712,38 @@ class DIType : public DIScope {
   DIFlags Flags;
   uint64_t SizeInBits;
   uint64_t OffsetInBits;
+  uint32_t NumExtraInhabitants;
```
Does the Swift frontend ever use the OffsetInBits? If not, we could store the extra inhabitants in there. (That would be an implementation detail of DIType and DIType would assert that you don't try to set both extra inhabitants and an offset).

adrian-prantl replied:
For everything else (Bitcode, Interface, assembler) leave it as it.

augusto2112 replied:
Swift uses `OffsetInBits` for `DIDerivedType` but not for `DIBasicType` or `DICompositeType`. `OffsetInBits` is used by other compilers in `DICompositeType` though. I prototyped what you suggested here but I think it'd make things significantly more complex and error prone for little gain. For example, when printing a `DICompositeType` we wouldn't be able to know if the field is an offset or the number of extra inhabitants, unless we add a flag tracking that on `DIType` (or tracking that in `DICompositeType` and making `getOffsetInBits` virtual) which I think would defeat the purpose of sharing the field for both types...

augusto2112 committed:
address comments

augusto2112 committed:
Fix test

zmodem commented on llvm/lib/IR/DIBuilder.cpp:
```diff
@@ -551,7 +555,7 @@ DIBuilder::createVariantPart(DIScope *Scope, StringRef Name, DIFile *File,
       VMContext, dwarf::DW_TAG_variant_part, Name, File, LineNumber,
       getNonCompileUnitScope(Scope), nullptr, SizeInBits, AlignInBits, 0, Flags,
-      Elements, 0, nullptr, nullptr, UniqueIdentifier, Discriminator);
+      Elements, 0, nullptr, nullptr, UniqueIdentifier, 0, Discriminator);
```
This looks wrong. `DICompositeType::get` takes the new `NumExtraInhabitants` parameter at the end. Your change is now passing 0 for Discriminator, Discriminator for the DataLocation parameter, etc. That broke rustc's test suite, see (parts of) https://crbug.com/377743159

Since `DICompositeType::get`'s `NumExtraInhabitants` was added at the end, and with a default value, no change here is necessary.

I've followed up in https://github.com/llvm/llvm-project/commit/546066e4f74d50c974248b0ed247f65ebf24b75c

augusto2112 replied:
Thanks @zmodem

Michael137 replied:
This happened [previously too](https://github.com/llvm/llvm-project/pull/72234#issuecomment-1813093440) with a similar change to the `::get` APIs. Wonder if it's worth doing something about this to catch these types of issues at compile time. E.g., using stronger types is my first instinct.
